### ISSUE TITLE
Virtual Block-Addressable Filesystem (WALFS)

### ISSUE THREAD

**Issue by WillEastbury** [WillEastbury/pios#10]
## What
Implement a minimal append-only WAL (Write-Ahead Log) based filesystem called **WALFS** that sits on top of our raw SD block driver. Provides three access modes: memory-mapped (MMAP-style), async FIFO write, and async FIFO scan/read. Tree-structured directories with Latin-1 filenames, Unix-like permissions, and crash-safe append-only semantics.

## Why
PIOS currently has raw block I/O but no way to organise data. Applications on user cores need to create, read, and persist named files. A WAL-based append-only design is ideal for our architecture because:
- **Crash-safe** — never overwrites existing data, only appends. Power loss = truncate last incomplete record.
- **Simple** — no free-space bitmap, no block allocator, no journaling complexity. Just a log.
- **Deterministic** — no garbage collection pauses, no defragmentation. Write performance is constant.
- **Multi-core friendly** — one writer core (Core 1/Disk), readers can scan the log independently via FIFO.

## Architecture

```
User Core (2/3)                    Core 1 (Disk)
      │                                  │
      ├─ MMAP: read file via            │
      │  virtual block window            │
      │  (flat memory region)            │
      │                                  │
      ├─ FIFO Write ──────────────→ WAL Appender ──→ SD Blocks
      │  (async, non-blocking)           │
      │                                  │
      ├─ FIFO Scan Request ──────→ Log Scanner ──→ Read SD
      │                                  │
      ←── FIFO Scan Result ──────── Returns records
```

### Superblock (Block 0, 512 bytes)
```c
struct walfs_super {
    u32 magic;            // 0x57414C46 ('WALF')
    u32 version;          // 1
    u32 block_size;       // 512
    u32 total_blocks;     // from SD card capacity
    u64 wal_head;         // next free byte offset in WAL region
    u64 tree_root;        // byte offset of current root inode record
    u32 record_count;     // total WAL records written
    u32 crc32;            // CRC32C of this superblock (hw accelerated)
    u8  label[32];        // volume label (Latin-1)
    u8  _reserved[448];
};
```

### WAL Region (Blocks 1+, append-only)
Everything after the superblock is a sequential append-only log of records:

```c
struct wal_record {
    u32 magic;            // 0x5245434F ('RECO')
    u32 type;             // RECORD_INODE, RECORD_DATA, RECORD_DIRENT, RECORD_DELETE
    u32 length;           // total record length including header
    u32 crc32;            // CRC32C of entire record (hw accelerated)
    u64 seq;              // monotonic sequence number
    u64 timestamp;        // cntvct_el0 at write time
    u8  payload[];        // type-specific data
};
```

### Record Types

**RECORD_INODE (type=1)** — file/directory metadata
```c
struct walfs_inode {
    u64 inode_id;         // unique ID (sequence number at creation)
    u64 parent_id;        // parent directory inode ID (0 = root)
    u32 mode;             // file type + permissions (rwxrwxrwx, 9 bits)
    u32 flags;            // WALFS_DIR=0x01, WALFS_FILE=0x02, WALFS_DELETED=0x80
    u64 size;             // file size in bytes (updated on each data append)
    u64 created;          // timestamp
    u64 modified;         // timestamp
    u8  name[128];        // Latin-1 encoded filename, null-terminated
};
```

**RECORD_DATA (type=2)** — file content chunk
```c
struct walfs_data {
    u64 inode_id;         // which file this data belongs to
    u64 offset;           // byte offset within file
    u32 length;           // data length (max 4096 per record)
    u8  data[];           // actual file content
};
```

**RECORD_DIRENT (type=3)** — directory entry (links name to inode)
```c
struct walfs_dirent {
    u64 parent_id;        // directory inode ID
    u64 child_id;         // child inode ID
    u8  name[128];        // Latin-1 name (redundant with inode, for fast scan)
};
```

**RECORD_DELETE (type=4)** — soft delete (marks inode as deleted)
```c
struct walfs_delete {
    u64 inode_id;         // inode to mark deleted
};
```

## Tree Structure

Directories form a tree via parent_id links. To list a directory, scan the WAL for the latest RECORD_DIRENT entries with matching parent_id. Deleted entries are filtered by checking for a subsequent RECORD_DELETE.

The **latest record for a given inode wins** — this is the append-only semantic. Updating a file's metadata means appending a new RECORD_INODE with the same inode_id. Reading file content means collecting all RECORD_DATA for that inode_id, ordered by offset.

### Mode 1: MMAP Virtual Block Window (~1KB)
1. Allocate a region in the user core's 16MB private RAM
2. Map a file's content into this region by reading RECORD_DATA chunks
3. Provide `walfs_mmap(inode_id, offset, length, dest_ptr)` — fills dest_ptr with file content
4. Read-only (writes go through FIFO)
5. Core 1 services the read via existing `MSG_DISK_READ` path

### Mode 2: Async FIFO Write (~800B)
New FIFO message types from user core → Core 1:
```c
MSG_FS_CREATE    = 30   // Create file/dir: payload = walfs_inode
MSG_FS_WRITE     = 31   // Append data: param=inode_id, buffer=data, length=len
MSG_FS_DELETE    = 32   // Delete: param=inode_id
MSG_FS_MKDIR     = 33   // Create directory
MSG_FS_DONE      = 34   // Ack from Core 1: status=0 ok
MSG_FS_ERROR     = 35   // Error from Core 1: status=error code
```
Core 1's disk loop appends the appropriate WAL record to the log and flushes to SD.

### Mode 3: Async FIFO Scan (~800B)
User core requests a directory listing or file read:
```c
MSG_FS_READDIR   = 40   // List directory: param=parent_inode_id
MSG_FS_READ      = 41   // Read file: param=inode_id, buffer=dest, length=max
MSG_FS_STAT      = 42   // Get inode metadata: param=inode_id
MSG_FS_FIND      = 43   // Find by name: buffer=path string
MSG_FS_SCAN_RESULT = 44 // Result: buffer=data, length=actual
```
Core 1 scans the WAL, resolves the latest state, and replies via FIFO.

## Latin-1 Filename Rules
- Max 127 bytes (128 with null terminator)
- Valid characters: 0x20-0x7E (printable ASCII) + 0xA0-0xFF (Latin-1 extended)
- Forbidden: 0x00-0x1F (control chars), 0x7F (DEL), 0x80-0x9F (C1 controls), '/' (0x2F, path separator)
- Case-sensitive
- No Unicode, no UTF-8 — pure Latin-1 (single-byte, deterministic comparison)

## Permissions Model
Simple Unix-style 9-bit mode: `rwxrwxrwx` (owner/group/other).
- No UID/GID — each core is implicitly a \"user\" (core 2 = user 0, core 3 = user 1)
- Check: core's position maps to owner (same core that created) or other
- Execute bit = allow core to treat file content as function pointer (dangerous but useful for code loading)

## CRC32 Integrity
Every record and the superblock are checksummed with hardware CRC32C (`hw_crc32c()` from simd.c). On read/scan, verify CRC before trusting record content. Corrupt records are skipped.

## Files to Create
- `include/walfs.h` — on-disk structures, FIFO message types, API
- `src/walfs.c` — WAL append, scan, tree resolution, MMAP helper

## Estimated Size
~3-4KB code, ~2KB RAM (index cache, scan buffers)

## Dependencies
Requires existing SD block driver (`sd.h`) and inter-core FIFO (`fifo.h`). Uses `hw_crc32c()` from `simd.h` for integrity checks.
